internal/services/elastic/elastic_test.go

```diff
@@ -0,0 +1,323 @@
+			excludedFields: []string{"data.misc.cell", "data.vehicle.currentLocation"},
+		},
+		{
+			name: "PrivilegeIDs set to Commands",
+			params: elastic.GetHistoryParams{
+				DeviceID:     "1",
+				StartTime:    testStartTime,
+				Buckets:      1,
+				EndTime:      testStartTime.Add(time.Millisecond * 1),
+				PrivilegeIDs: []int64{constants.Commands},
+			},
+			expectedIds:    []string{"0"},
+			expectedFields: []string{"data.vehicle.powertrain"},
+			excludedFields: []string{"data.misc.cell", "data.vehicle.currentLocation"},
+		},
+		{
+			name: "PrivilegeIDs set to AllTimeLocation and NonLocationData",
+			params: elastic.GetHistoryParams{
+				DeviceID:     "1",
+				StartTime:    testStartTime,
+				Buckets:      1,
+				EndTime:      testStartTime.Add(time.Millisecond * 1),
+				PrivilegeIDs: []int64{constants.NonLocationData, constants.AllTimeLocation},
+			},
+			expectedIds:    []string{"0"},
+			expectedFields: []string{"data.vehicle.powertrain"},
+			excludedFields: []string{"data.misc.cell", "data.vehicle.currentLocation"},
+		},
+		{
+			name: "PrivilegeIDs set to AllTimeLocation and Commands",
+			params: elastic.GetHistoryParams{
+				DeviceID:     "1",
+				StartTime:    testStartTime,
+				Buckets:      1,
+				EndTime:      testStartTime.Add(time.Millisecond * 1),
+				PrivilegeIDs: []int64{constants.Commands, constants.AllTimeLocation},
+			},
+			expectedIds:    []string{"0"},
+			expectedFields: []string{"data.vehicle.powertrain", "data.vehicle.currentLocation", "data.misc.cell"},
+		},
+	}
+
+	// For each test case, call the GetHistory function with the custom parameters and check if the result is equal to the expected result.
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := service.GetHistory(context.Background(), tc.params)
+			if err != nil {
+				require.NoErrorf(t, err, "could not get history: %v", err)
+			}
+			// check if each id in the result is in the expected ids
+			ids := []string{}
+			for _, hit := range result {
+				hitMap, ok := hit.(map[string]any)
+				require.Truef(t, ok, "results is not a list of map instead a list of %T", hit)
+				idVal, ok := hitMap["testID"]
+				require.Truef(t, ok, "testID field not found in result")
+				id, ok := idVal.(string)
+				require.Truef(t, ok, "testID field not a string")
+				ids = append(ids, id)
+				hitData, err := json.Marshal(hit)
+				require.NoErrorf(t, err, "could not marshal hit: %v", err)
+				// verify that the expected fields are present in the result
+				for _, field := range tc.expectedFields {
+					fieldData := gjson.GetBytes(hitData, field)
+					require.Truef(t, fieldData.Exists(), "expected field %q not found in result", field)
+				}
+				for _, field := range tc.excludedFields {
+					fieldData := gjson.GetBytes(hitData, field)
+					require.Falsef(t, fieldData.Exists(), "excluded field %q found in result", field)
+				}
+			}
+			slices.Sort(tc.expectedIds)
```

Comment:
We do kinda care about the order though, yeah?

Reply:
Since we do not use the [BucketSort](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-pipeline-bucket-sort-aggregation.html#search-aggregations-pipeline-bucket-sort-aggregation) Aggregation we can't guarantee the order. Though I do think it would be beneficial to sort by time.

Reply:
From [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html#date-histogram-order),

> By default the returned buckets are sorted by their key ascending, but you can control the order using the order setting. This setting supports the same order functionality as [Terms Aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#search-aggregations-bucket-terms-aggregation-order).